versioned/persist/adapter/src/main/java/org/projectnessie/versioned/persist/adapter/ReferencesUtil.java

```diff
@@ -0,0 +1,192 @@
+  }
+
+  /**
+   * Identifies all heads and fork-points.
+   *
+   * <ul>
+   *   <li>"Heads" are commits that are not referenced by other commits.
+   *   <li>"Fork points" are commits that are the parent of more than one other commit. Knowing
+   *       these commits can help to optimize the traversal of commit logs of multiple heads.
+   * </ul>
+   *
+   * @param expectedCommitCount it is recommended to tell the implementation the total number of
+   *     commits in the Nessie repository
+   */
+  public HeadsAndForkPoints identifyAllHeadsAndForkPoints(int expectedCommitCount) {
+    // Using open-addressing implementation here, because it's much more space-efficient than
+    // java.util.HashSet.
+    Set<Hash> parents = newOpenAddressingHashSet(expectedCommitCount);
+    Set<Hash> heads = newOpenAddressingHashSet();
+    Set<Hash> forkPoints = newOpenAddressingHashSet();
+
+    // Need to remember the time when the identification started, so that a follow-up
+    // identifyReferencedAndUnreferencedHeads() knows when it can stop scanning a named-reference's
+    // commit-log. identifyReferencedAndUnreferencedHeads() has to read up to the first commit
+    // _before_ this timestamp to not commit-IDs as "unreferenced".
+    //
+    // Note: keep in mind, that scanAllCommitLogEntries() returns all commits in a
+    // non-deterministic order. Example: if (at least) two commits are added to a branch while this
+    // function is running, the original HEAD of that branch could otherwise be returned as
+    // "unreferenced".
+    long scanStartedAtInMicros = databaseAdapter.commitTimeInMicros();
+
+    // scanAllCommitLogEntries() returns all commits in no specific order, parents may be scanned
+    // before or after their children.
+    try (Stream<CommitLogEntry> scan = databaseAdapter.scanAllCommitLogEntries()) {
+      scan.forEach(
+          entry -> {
+            Hash parent = entry.getParents().get(0);
+            if (!parents.add(parent)) {
+              // If "parent" has already been added to the set of parents, then it must be a
+              // fork point.
+              forkPoints.add(parent);
+            } else {
+              // Commits in "parents" that are also contained in "heads" cannot be HEADs.
+              // This can happen because the commits are scanned in "random order".
+              // Should do this here to prevent the "heads" set from becoming unnecessarily big.
+              heads.remove(parent);
+            }
+
+            Hash commitId = entry.getHash();
+            if (!parents.contains(commitId)) {
+              // If the commit-ID is not present in "parents", it must be a HEAD
+              heads.add(commitId);
+            }
+          });
+    }
+
+    // Commits in "parents" that are also contained in "heads" cannot be HEADs.
+    // This can happen because the commits are scanned in "random order".
+    heads.removeIf(parents::contains);
```

Comment:
Is it really possible to get false-positive heads at this point? We remove each new parent from `heads` and we only add to `heads` if the hash is not in `parents` (above)... it looks like false-positive are not possible.

Reply:
Yea - I think the same (because it's handled in the loop above), because I stepped through here in the debugger and this removeIf never did anything.

Just not sure whether there's maybe a case where it could be. I can add a TODO to visit again. WDYT?

Reply:
Yeah, a TODO is fine. The number of heads is not that great anyway.

Reply:
I suppose we can have a test where we purposefully execute all input permutations. That should validate the head identification logic even with a small dataset. I guess we really need a few liner segments and a few fork points (with nesting).